packages/openneuro-server/src/graphql/resolvers/dataset-search.js

```diff
@@ -103,3 +105,89 @@ export default {
     },
   },
 }
+
+const addClause = (query, type, clause) => {
+  if (query.bool[type]) {
+    query.bool[type] = [...query.bool[type], clause]
+  } else {
+    query.bool[type] = [clause]
+  }
+}
+
+const parseQuery = async (query, datasetType, datasetStatus, userId) => {
+  const queryObj = JSON.parse(query)
+  if (datasetType === 'Following') {
+    const results = await Subscription.find({ userId })
+    const followedDatasets = results.map(({ datasetId }) => datasetId)
+    const termsClause = {
+      terms: {
+        id: followedDatasets,
+      },
+    }
+    addClause(queryObj, 'filter', termsClause)
+  } else if (datasetType === 'My Bookmarks') {
+    const results = await Star.find({ userId })
+    const bookmarkedDatasets = results.map(({ datasetId }) => datasetId)
+    const termsClause = {
+      terms: {
+        id: bookmarkedDatasets,
+      },
+    }
+    addClause(queryObj, 'filter', termsClause)
+  } else if (datasetType === 'My Datasets') {
+  }
+  return queryObj
+}
+
+/**
+ * Search result cursor resolver
+ * TODO this is a Relay pagination type and could use the interface
+ * @param {any} obj
+ * @param {object} args
+ * @param {string} args.query Stringified Query (DSL) argument for ElasticSearch
+ * @param {string} args.datasetType Stringified Query (DSL) argument for ElasticSearch
+ * @param {string} args.datasetStatus Stringified Query (DSL) argument for ElasticSearch
+ * @param {string} args.sortBy Stringified Query (DSL) argument for ElasticSearch
+ * @param {string} args.after Cursor for paging forward
+ * @param {number} args.first Limit of entries to find
+ */
+export const advancedDatasetSearchConnection = async (
+  obj,
+  { query, datasetType, datasetStatus, sortBy, after, first = 25 },
```

Comment:
For these JSON types, I think we should use something like the JSON scalar from [graphql-scalars](https://github.com/Urigo/graphql-scalars#readme). It would make this API a lot more precise than string without needing to specify the whole ElasticSearch type.

Reply:
Ended up using the JSON scalar provided in graphql-compose.